google-cloud-storage/src/main/java/com/google/cloud/storage/DefaultBufferedReadableByteChannel.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import com.google.cloud.storage.BufferedReadableByteChannelSession.BufferedReadableByteChannel;
+import com.google.cloud.storage.UnbufferedReadableByteChannelSession.UnbufferedReadableByteChannel;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+
+final class DefaultBufferedReadableByteChannel implements BufferedReadableByteChannel {
+
+  private final ByteBuffer buffer;
+
+  private final UnbufferedReadableByteChannel channel;
+
+  DefaultBufferedReadableByteChannel(ByteBuffer buffer, UnbufferedReadableByteChannel channel) {
+    this.buffer = buffer;
+    this.channel = channel;
+  }
+
+  @Override
+  public boolean isComplete() {
+    return channel.isComplete();
+  }
+
+  long totalRead = 0;
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    if (!channel.isOpen()) {
+      throw new ClosedChannelException();
+    }
+
+    int bytesConsumed = 0;
+
+    while (dst.hasRemaining()) {
+      int bufferRemaining = buffer.remaining();
+
+      int dstRemaining = dst.remaining();
+      int dstPosition = dst.position();
+
+      final int tmpBytesCopied;
+      boolean enqueuedBytes = enqueuedBytes();
+      if (enqueuedBytes) {
```

Comment:
Will do. Though, having the variable defined means that the value shows up in the debugger if the breakpoint is set at line 59 here, rather than having to add a separate watch.